torch_geometric/contrib/nn/models/hin2vec.py

```diff
@@ -0,0 +1,375 @@
+
+    def loader(self, **kwargs):
+        r"""Returns the data loader that creates both positive and negative
+        random walks on the heterogeneous graph.
+
+        Args:
+            **kwargs (optional): Arguments of
+                :class:`torch.utils.data.DataLoader`, such as
+                :obj:`batch_size`, :obj:`shuffle`, :obj:`drop_last` or
+                :obj:`num_workers`.
+        """
+        return DataLoader(range(self.node_count * self.walks_per_node),
+                          collate_fn=self._sample, **kwargs)
+
+    def _sample(self, batch: Union[List[int],
+                                   Tensor]) -> Tuple[Tensor, Tensor]:
+        r"""Generate positive and negative samples within self.metapath_length
+        from the random walk with each row of sample of form
+        (start_node end_node metapath_id).
+        """
+        if not isinstance(batch, Tensor):
+            batch = torch.tensor(batch, dtype=torch.long)
+
+        node_seq, edge_seq = self.rw[0][batch], self.rw[1][batch]
+
+        pos = []
+        for w in range(self.metapath_length):
+            for i in range(edge_seq.size(1) - w):
+                metapath = edge_seq[:, i:i + w + 1]
+                mask = metapath >= 0  # Filter end of walk
+                mask = mask.all(1)
+                metapath = metapath[mask, :]
+                # Match metapath to id
+                metapath_id = (metapath.unsqueeze(1) == self.metapath_dict[w +
+                                                                           1]
+                               ).all(-1).nonzero()[:, -1]
+                assert metapath_id.size(0) == metapath.size(0)
+                metapath_id += self.path_start[w + 1]
+
+                start_nodes = node_seq[:, i][mask]
+                end_nodes = node_seq[:, i + w + 1][mask]
+                pos.append(
+                    torch.stack((start_nodes, end_nodes, metapath_id), dim=-1))
+        pos_sample = torch.cat(pos, dim=0)
+        if not self.allow_circle:
+            pos_sample = self._remove_circle(pos_sample)
+
+        # Corrupt the tail to generate negative sample
+        neg_sample = pos_sample.repeat(self.num_negative_samples, 1)
+        if self.same_neg_node_type:
+            neg_node_idx = torch.rand((neg_sample.size(0), ))
+            # Multiply with count of same node type as positive sample
+            neg_node_idx = torch.floor(
+                neg_node_idx * self.node_type_table[neg_sample[:, 1], 0])
+            neg_node_idx += self.node_type_table[neg_sample[:, 1], 1]
+            neg_sample[:, 1] = neg_node_idx.long()
+        else:
+            neg_sample[:, 1] = torch.randint(0, self.node_count - 1,
+                                             (neg_sample.size(0), ))
+        if not self.allow_circle:
+            neg_sample = self._remove_circle(neg_sample)
+
+        return pos_sample, neg_sample
+
+    def _remove_circle(self, sample: Tensor) -> Tensor:
```

Comment:
cycle instead of circle?
```suggestion
    def _remove_cycle(self, sample: Tensor) -> Tensor:
```

Reply:
Updated to cycle